File: doc/code/cpp/assembler.cpp
#include <cstdint>
#include <memory>
#include <string>
#include <vector>

#include <LIEF/Abstract/Binary.hpp>
#include <LIEF/ELF.hpp>
#include <LIEF/asm/AssemblerConfig.hpp>
#include <LIEF/asm/Instruction.hpp>
#include <optional>

void disassemble_assemble() {
  // lief-doc: disassemble-assemble-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  std::vector<uint64_t> syscall_addresses;
  for (const auto& inst : elf->disassemble(0x400090)) {
    if (inst.is_syscall()) {
      syscall_addresses.push_back(inst.address());
    }
  }

  for (uint64_t addr : syscall_addresses) {
    elf->assemble(addr, R"asm(
      mov x1, x0;
      str x1, [x2, #8];
    )asm");
  }
  // lief-doc: disassemble-assemble-end
}

void context_error() {
  // lief-doc: context-error-start
  std::unique_ptr<LIEF::ELF::Binary> elf;
  elf->assemble(elf->entrypoint(), R"asm(
    mov rdi, rax;
    call a_custom_function;
  )asm");
  // lief-doc: context-error-end
}

void config_resolver() {
  // lief-doc: config-resolver-start
  class MyConfig : public LIEF::assembly::AssemblerConfig {
    public:
    std::optional<uint64_t> resolve_symbol(const std::string& name) override {
      if (name == "a_custom_function") {
        return 0x1000;
      }
      return std::nullopt;
    }
  };

  MyConfig myconfig;

  std::unique_ptr<LIEF::ELF::Binary> elf;

  elf->assemble(elf->entrypoint(), R"asm(
    mov rdi, rax;
    call a_custom_function;
  )asm",
                myconfig);
  // lief-doc: config-resolver-end
}

void config_target() {
  // lief-doc: config-target-start
  class MyConfig : public LIEF::assembly::AssemblerConfig {
    public:
    MyConfig() = delete;
    MyConfig(LIEF::Binary& target) :
      LIEF::assembly::AssemblerConfig(),
      target_(&target) {}

    std::optional<uint64_t> resolve_symbol(const std::string& name) override {
      if (auto addr = target_->get_function_address(name)) {
        return *addr;
      }
      return std::nullopt;
    }

    ~MyConfig() override = default;

    private:
    LIEF::Binary* target_ = nullptr;
  };

  std::unique_ptr<LIEF::ELF::Binary> elf;
  MyConfig myconfig(*elf);

  elf->assemble(elf->entrypoint(), R"asm(
    mov rdi, rax;
    call a_custom_function;
  )asm",
                myconfig);
  // lief-doc: config-target-end
}

void disable_instruction() {
  // lief-doc: nop-out-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  // Overwrite the first call in the region (e.g. a call to an anti-debugging
  // routine) with nops
  for (const auto& inst : elf->disassemble(0x401200)) {
    if (inst.is_call()) {
      std::string nops;
      for (size_t i = 0; i < inst.size(); ++i) {
        nops += "nop\n";
      }
      elf->assemble(inst.address(), nops);
      break;
    }
  }

  elf->write("patched.bin");
  // lief-doc: nop-out-end
}

File: doc/code/python/assembler.py
from typing import override

import lief


def disassemble_assemble(elf: lief.ELF.Binary) -> None:
    # lief-doc: disassemble-assemble-start
    elf: lief.ELF.Binary

    syscall_addresses = [
        inst.address for inst in elf.disassemble(0x400090) if inst.is_syscall
    ]

    for syscall_addr in syscall_addresses:
        elf.assemble(
            syscall_addr,
            """
            mov x1, x0;
            str x1, [x2, #8];
            """,
        )
    # lief-doc: disassemble-assemble-end


def context_error(elf: lief.ELF.Binary) -> None:
    # lief-doc: context-error-start
    elf: lief.ELF.Binary

    elf.assemble(
        elf.entrypoint,
        """
        mov rdi, rax;
        call a_custom_function;
        """,
    )
    # lief-doc: context-error-end


def config_resolver(elf: lief.ELF.Binary) -> None:
    # lief-doc: config-resolver-start
    class MyConfig(lief.assembly.AssemblerConfig):
        def __init__(self):
            super().__init__()  # Important!

        @override
        def resolve_symbol(self, name: str) -> int | None:
            if name == "a_custom_function":
                return 0x1000
            return None

    elf: lief.ELF.Binary

    elf.assemble(
        elf.entrypoint,
        """
        mov rdi, rax;
        call a_custom_function;
        """,
        MyConfig(),
    )
    # lief-doc: config-resolver-end


def config_target(elf: lief.ELF.Binary) -> None:
    # lief-doc: config-target-start
    class MyConfig(lief.assembly.AssemblerConfig):
        def __init__(self, target: lief.Binary):
            super().__init__()  # Important!

            self._target = target

        @override
        def resolve_symbol(self, name: str) -> int | None:
            addr = self._target.get_function_address(name)
            if isinstance(addr, lief.lief_errors):
                return None
            return addr

    elf: lief.ELF.Binary
    config = MyConfig(elf)

    elf.assemble(
        elf.entrypoint,
        """
        mov rdi, rax;
        call a_custom_function;
        """,
        config,
    )
    # lief-doc: config-target-end


def disable_instruction(elf: lief.ELF.Binary) -> None:
    # lief-doc: nop-out-start
    elf: lief.ELF.Binary

    # Overwrite the first call in the region (e.g. a call to an anti-debugging
    # routine) with nops
    for inst in elf.disassemble(0x401200):
        if inst.is_call:
            elf.assemble(inst.address, "nop\n" * inst.size)
            break

    elf.write("patched.bin")
    # lief-doc: nop-out-end
